### DISCUSSION TITLE
Safer handling of internal soname dependencies

### DISCUSSION THREAD

**Discussion by pmatilai** [rpm-software-management/rpm#2872] [Ideas]
This is one of those topics that gets raised semi-annually at least. To point out a couple, https://bugzilla.redhat.com/show_bug.cgi?id=2259260 and https://bugzilla.redhat.com/show_bug.cgi?id=1464368 and https://github.com/rpm-software-management/rpm/issues/1227 but there are certainly more out there.

For some cases like an unfortunately named project starting with lib having a Python/Perl module, the provide may be mostly harmless if ugly. The far worse issue is bundled libraries that are incompatible with the system ones and not even in the path, but create identical provide names which a depsolver may end up pulling instead of the system level library, and break things quite badly.

The fundamental issue may well be unsolvable, the main focus here is to look for a solution that makes those bundled libraries less dangerous. If other problems get solved while we're at it, great. 

The common suggestion is to filter out soname dependencies outside the system library path, but this has multiple problems: filtering them on the rpm metadata level doesn't make the dependencies go away, and they may well be needed for sub-packages and so on. What may work better is transforming the sonames dependencies for libraries outside system path to something else, perhaps including the parent package name in the name so it can't be mixed with the system or somebody elses bundled library. For provides one can do this by parsing /etc/ld.so.conf[.d] but requires seems to require solving, which is something elfdeps doesn't (and IMO, shouldn't) do. (I don't like parsing other people's config files either, but that seems unavoidable)

One simple possibility would be giving elfdeps a list/regex of soname prefixes to match. Eg, passing "libxml2" would cause it to map any dependencies found on libxml* to some other, internal-suggesting name. Such as %{name}-libxml*. That list/regex could be set by the packager, but to satisfy the goal of bundled library safety, it needs to happen automatically for provides. I guess something like this would do:

1. figure out system library path, considering ld.so.conf[.d] in the buildroot and everything
2. find out any shared libraries in the buildroot outside that path
3. set the elfdeps mangler (eg %{name}-<dependencyz>) to apply to everything found in 2
4. profit?

Mangling discovered dependencies isn't hard, retrofitting the discovery parts to the rpm framerwork probably is.

**Comment by codonell** [comment]
@pmatilai Can you expand a bit on what you mean by not liking parsing other people's config files? Is it that you would like a clean way to discover the system library paths searched by the dynamic loader?

**Reply by praiskup** [reply]
The last time I observed (few years ago in 1464368) the tooling did not seem to exist. Would you mind providing such a tool in glibc-devel?

**Reply by pmatilai** [reply]
Yes, a clean way to discover the system library paths searched by the dynamic loader would go a long way.

Hmm. I guess this is kinda doable by running `ldconfig -r <buildroot>` followed by `ldconfig -r <buildroot> -p` from which you can pull the directories (so effectively the path) or the files themselves. Just do that with an rpm-controlled cache file that can then be removed from the buildroot. Ability to use existing tools would be a big bonus.

The tricky part is handling /etc/ld.so.conf.d parts from the buildroot - I assume ldconfig wont look into that unless instructed by the main config specifically, and for that we'd need to also use a private ld.so.conf copy from the system inside the root.

Edit: ...and then it needs to pull the full host configuration *in addition* to the one inside the buildroot, because the thing in the buildroot could be installing to one of the locations included in system /etc/ld.so.conf.d.

**Reply by pmatilai** [reply]
The above *almost* works. It's just that the ld.so.conf from Fedora causes an "ldconfig: need absolute file name for configuration file when using -r" error from ldconfig. Had me scratching hair for a bit because I thought it ws complaining about the path I was passing in -f which *was* absolute.

This means we can't just use the copy from the system, we'd need to use a different one which hardcodes /etc/ld.so.conf.d as the include path. I know /etc/ld.so.conf doesn't change a lot in practise, but having to write out a hardcoded path means we could just as well just hardwire the path and parse those files by ourselves, it's not like the format is difficult. Probably less work than trying to abuse ldconfig to give up its secrets :smile:

**Reply by codonell** [reply]
When we talk about 1464368 I assume we mean https://bugzilla.redhat.com/show_bug.cgi?id=1464368

@praiskup When you say "tooling" I expect that between us we need to decide what we need and why we need it (to document the reason for the feature). Is there a specific tool you wish you had? What would that tool do?

@pmatilai My opinion here is that `ldconfig` and in fact any path printing based on the current configuration won't help. Here is why. The set of available runtime paths is only visible when you install the package and all the required dependencies and this isn't what we have in the buildroot. As an example the package is allowed to install a library to a known hardware path e.g. `/usr/lib64/glibc-hwcaps/x86-64-v4` and then document that it must be installed only on hardware that supports x86-64-v4 or newer. If the builder is x86-64-v3 then this path is never searched and never printed. We would exclude that DSO from the search and never generate the automatic provides for it. I think this is actually the *correct* thing to do, that outside of the default fallback systems paths e.g. `/lib64` and `/usr/lib64` there just isn't a correct way to handle the system composition other than throwing everything in and letting the system fail spectacularly on occasion (like we see in the bug reports) or ignore everything not in `/lib64` and `/usr/lib64` (or the default paths as printed by `/usr/bin/ld.so --help`) and force packages to explicitly define what they bring to the system composition.

I expect that in general we have consensus that a solution that requires non-default path installs to specify all provided DSOs it not a useful tradeoff against occassional system composition failure.

Is there an alternative solution? For example, if a packaged rpm has a DSO... then that DSO should only get added to the automatically computed provides if there is a `/etc/ld.so.conf.d` fragment in the same rpm that specifies the path to that DSO? This works by argument that the rpm that provides the DSO also installs the conf fragment, which might not be the case, but I wonder if it's sufficient to resolve the majority of the cases.

**Reply by codonell** [reply]
@pmatilai I realize that my suggestion looks a lot like what you were attempting to do on a per-package basis? Were you trying to (a) create a template /etc/ld.so.conf which just includes the the fragment directory, and for the buildroot of the rpm build (b) allow it to parse any fragments that were created by the package and then print the list? That should work, or we can enhance `ldconfig` to make it work.

**Reply by pmatilai** [reply]
Yes, that possible ld.so.conf fragment in the buildroot needs to be covered. But, I think the hosts ld.so.conf[.d] needs to be processed as well. The currently building package could be installing to a subdirectory that some other package (which would have to be a build-dependency) has declared a system path by installing such a fragment.

At the baseline I need a way to determine whether a library in a package we're building is considered a system library or not, one that matches with how the linker sees the world. hwcaps paths are another quirk in this, but since they apparently are not in ld.so.conf[.d] paths, they would be not considered system libraries.

**Reply by pmatilai** [reply]
Note that I'm specifically avoiding talking about filtering these dependencies out because I think in many cases its more useful to turn them into something else instead.

**Reply by codonell** [reply]
@pmatilai Just for clarity we are leaving out `DT_RPATH`, environmental combination of `LD_LIBRARY_PATH` and the search paths (packages that alter shell defaults and global search paths), `LD_PRELOAD` interposition (similar to `LD_LIBRARY_PATH` but early loads the SONAME), and `DT_RUNPATH` and their effect on the resolution of the absolute path used to load the DSO *requried* by either an executable or library. These are AFAIK details that RPM's `elfdeps` does not model and so still requires the developer to express them directly as manual requires or provides. This is not a regression in behaviour just a modelling choice. After all our intent is to make this easier for developers without needing to expose all of the details around the ELF standard and the current Linux implementations.

Yes, I agree that the host (or mock chroot) ld.so.conf[.d] paths should be searched for requires. The still broken case is when the rpm installs a DSO into a path created by an rpm in the `Requires:` (rather than `BuildRequire:`) since this cannot be discovered at build time? In this case the rpm neither provides the ld.so.conf fragment, nor is the DSO used at build time. Is this a narrow enough case that we can ignore it? Initially my answer is: Yes, I think we should focus solely on the host ld.so.conf[.d] paths and the paths provided by the package itself (as it is being built)

It seems natural to use ldconfig to provide these paths? We could add a new option to just print directories searched.

**Reply by pmatilai** [reply]
Yeah elfdeps basically ignores all path-aspects, it only really handles DT_NEEDED and DT_SONAME.
One thing here though: a manually added require/provide is not equal to one discovered by rpm, the latter are tracked on file-level and rpm actually uses this information. That's why I'm so reluctant to filter out anything and rather turn them into something non-harmful instead (eg prefix discovered provides/requires by package name or such). And having packagers form those wacko ()(64bit) markers manually doesn't really fly either. 

The paths from the config is actually what I've been after all along, I hadn't really even considered there were other paths at play (silly me, of course the are). But yeah having an ldconfig option to print out the configured paths would be nice. If we can get it to print out the host config directories, we can use that to create a temporary config file to use inside the buildroot to include any fragments from there and print again to get the final list of paths for a package.

As for the Requires vs Buildrequires matter, I'd expect the vast majority of these cases to be covered by the foo-devel buildrequire to pull in the part with the ld.so.conf too. The corner-case where that doesn't happen (it's possible of course) can be worked around by just adding a Buildrequire on the thing with the ld.so.conf.d fragment, easy enough.

**Reply by pmatilai** [reply]
Actually, we'll need to have this path-discovery wrapped in some libc specific helper script in rpm because this eg musl has an entirely different library search path mechnism. That's fine though, for glibc we could implement our own ld.so.conf parser for the immediate use and then once the ldconfig native version is widely available, switch to that for glibc.

**Reply by pmatilai** [reply]
One thing just for the record, up to now elfdeps has been path agnostic by design to match the dynamic linker: it doesn't matter where those libraries come from as long as they're in the runtime path. Of course it's not an "end user tool" but having consistent output regardless of the file location is nice really. So I need to think/investingate whether the path dependent parts wouldn't be better done some place else, such as a separate filter/transform post step.

**Reply by codonell** [reply]
@pmatilai I agree that while a "search path" is a fundamental part of the model for the full view of the dynamic loader, it need not be part of rpm's `elfdeps` and could become some kind of filter option that is driven by libc specific search information.

**Comment by praiskup** [comment]
Seems related https://bugzilla.redhat.com/show_bug.cgi?id=2266553

**Reply by codonell** [reply]
@praiskup It is exactly the same problem. We will continue to have these issues at the distribution level until we change the way we model how a package exports an interface e.g. list of shared objects. There is perhaps a broader question that this mistake would *not* have been caught without the current semantics we have, but the flip side is that it fails so badly that it blocks work being done by others and is disruptive. While the failure would have been less spectacular we could have caught this in CI gating by detecting that this package was doing something wrong.

**Reply by praiskup** [reply]
Agreed, especially with the [CI gating part](https://matrix.to/#/!IHsypFkurbRbXzkjTO:matrix.org/$Jo1uzuG9SVrm__zmNvoYXXI2EarBYgIt-WdTu8TPBVg?via=fedora.im&via=matrix.org&via=fedoraproject.org).
